File: crates/core/tests/fixtures/dataframe_records_contract.py
"""Freeze typed record inference against the unchanged Qlib append function.

This is source characterization, not evidence of native record-constructor parity.
"""
import ast
import hashlib
import itertools
import json
import math
import sys
import warnings
from pathlib import Path

import numpy as np
import pandas as pd

source = Path(sys.argv[1]).read_bytes()
source_hash = hashlib.sha256(source).hexdigest()
assert source_hash == "89267f5cfc9e38751cb2c3a37c74ca712e8c395f492d53ed02f029a66272074f"
node = next(n for n in ast.parse(source).body if isinstance(n, ast.FunctionDef) and n.name == "dataframe_append")
future = ast.ImportFrom(module="__future__", names=[ast.alias(name="annotations")], level=0)
exec(compile(ast.fix_missing_locations(ast.Module(body=[future, node], type_ignores=[])), "source", "exec"))

# Reuse only the explicit lossless snapshot helpers, without executing the other matrix.
helpers = ast.parse(Path(__file__).with_name("dataframe_constructor_contract.py").read_bytes())
helpers = [n for n in helpers.body if isinstance(n, ast.FunctionDef) and n.name in {"scalar", "snapshot"}]
assert len(helpers) == 2
exec(compile(ast.Module(body=helpers, type_ignores=[]), "snapshot_helpers", "exec"))

samples = {"none": None, "na": pd.NA, "nat": pd.NaT, "bool": True,
           "integer": -3, "unsigned": 2**64-1, "bigint": 2**80,
           "float": 1.25, "negative_zero": -0.0, "nan": np.nan,
           "text": "中\ud800", "complex": 1+2j}
for dtype in ["int8", "int16", "int32", "int64", "uint8", "uint16", "uint32", "uint64",
              "float16", "float32", "float64", "complex64", "complex128", "bool"]:
    samples["numpy_" + dtype] = np.array([1], dtype=dtype)[0]
for unit in ["s", "ms", "us", "ns"]:
    for zone in [None, "UTC", "Asia/Shanghai"]:
        samples[f"timestamp_{unit}_{zone}"] = pd.Timestamp("2024-01-02", tz=zone).as_unit(unit)
    samples["duration_" + unit] = pd.Timedelta("1h").as_unit(unit)
samples["timestamp_outside_ns"] = pd.Timestamp("2500-01-01").as_unit("s")
samples["duration_outside_ns"] = pd.Timedelta(np.timedelta64(10**12, "s"))

lefts = {"empty": pd.DataFrame(),
         "initialized": pd.DataFrame(columns=["datetime", "x", "stock_id"]).set_index("datetime"),
         "populated": pd.DataFrame({"x": [0.5]}, index=pd.DatetimeIndex(["2024-01-01"]))}
cases = []
for (name, value), missing, order in itertools.product(samples.items(),
                                                     ["single", "omitted", "none", "nan", "nat", "na"],
                                                     ["forward", "reverse"]):
    first = {"x": value, "datetime": pd.Timestamp("2024-01-02")}
    second = {"stock_id": "SH600000", "datetime": pd.Timestamp("2024-01-03")}
    if missing not in {"single", "omitted"}:
        second["x"] = {"none": None, "nan": np.nan, "nat": pd.NaT, "na": pd.NA}[missing]
    records = [first] if missing == "single" else [first, second]
    if order == "reverse":
        records.reverse()
    before = [[(key, scalar(item)) for key, item in record.items()] for record in records]
    for left_name, left in lefts.items():
        left_before = snapshot(left)
        with warnings.catch_warnings(record=True) as captured:
            warnings.simplefilter("always")
            constructed = snapshot(pd.DataFrame(records))
        constructor_warnings = [[type(w.message).__name__, str(w.message)] for w in captured]
        with warnings.catch_warnings(record=True) as captured:
            warnings.simplefilter("always")
            try:
                outcome = {"output": snapshot(dataframe_append(left, records))}
            except Exception as error:
                outcome = {"error": type(error).__name__, "message": str(error)}
        assert snapshot(left) == left_before
        assert [[(key, scalar(item)) for key, item in record.items()] for record in records] == before
        cases.append(dict(name=name, missing=missing, order=order, left=left_name,
                          construction=constructed, construction_warnings=constructor_warnings,
                          warnings=[[type(w.message).__name__, str(w.message)] for w in captured], **outcome))

digest = hashlib.sha256(json.dumps(cases, sort_keys=True, separators=(",", ":")).encode()).hexdigest()
print(json.dumps(dict(source_sha256=source_hash, pandas=pd.__version__, numpy=np.__version__,
                      samples=len(samples), digest=digest, cases=cases)))

File: crates/core/tests/dataframe_records_contract.rs
//! Source characterization only; native typed-record inference is still pending.
use serde_json::{Value, json};
use std::{path::PathBuf, process::Command};

#[test]
fn records_reinfer_typed_scalars_and_preserve_discovery_order() {
    let root = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let output = Command::new("python")
        .arg(root.join("tests/fixtures/dataframe_records_contract.py"))
        .arg(root.join("../../../qlib/qlib/rl/order_execution/utils.py"))
        .output()
        .unwrap();
    assert!(
        output.status.success(),
        "{}",
        String::from_utf8_lossy(&output.stderr)
    );
    let contract: Value = serde_json::from_slice(&output.stdout).unwrap();
    assert_eq!(contract["pandas"], "2.3.3");
    assert_eq!(contract["numpy"], "2.4.0");
    assert_eq!(contract["samples"], 44);
    assert_eq!(
        contract["digest"],
        "ea9d83e7e17d1d3ed5400f48ed58a1ca353475ce01474a34b7e6721c49204ae1"
    );
    let cases = contract["cases"].as_array().unwrap();
    assert_eq!(cases.len(), 1584);
    assert!(cases.iter().all(|c| c.get("output").is_some()));
    assert!(
        cases
            .iter()
            .all(|c| c["construction_warnings"] == json!([]))
    );
    let find = |name: &str, missing: &str, order: &str| {
        &cases
            .iter()
            .find(|c| {
                c["name"] == name
                    && c["missing"] == missing
                    && c["order"] == order
                    && c["left"] == "empty"
            })
            .unwrap()["construction"]
    };
    for (name, dtype) in [("numpy_int8", "int8"), ("numpy_float32", "float32")] {
        assert_eq!(find(name, "single", "forward")["dtypes"][0], dtype);
        for missing in ["omitted", "none", "nan"] {
            assert_eq!(find(name, missing, "forward")["dtypes"][0], "float64");
        }
        for missing in ["nat", "na"] {
            assert_eq!(find(name, missing, "forward")["dtypes"][0], "object");
        }
    }
    for (name, dtype) in [
        ("timestamp_s_None", "datetime64[ns]"),
        ("timestamp_s_UTC", "datetime64[ns, UTC]"),
        ("timestamp_s_Asia/Shanghai", "datetime64[ns, Asia/Shanghai]"),
        ("duration_s", "timedelta64[ns]"),
    ] {
        for missing in ["single", "omitted", "none", "nan", "nat"] {
            assert_eq!(find(name, missing, "forward")["dtypes"][0], dtype);
        }
        assert_eq!(find(name, "na", "forward")["dtypes"][0], "object");
    }
    for name in ["timestamp_outside_ns", "duration_outside_ns"] {
        assert_eq!(find(name, "single", "forward")["dtypes"][0], "object");
        assert_eq!(
            find(name, "single", "forward")["values"][0][0][1],
            if name == "timestamp_outside_ns" {
                "datetime64[s]"
            } else {
                "timedelta64[s]"
            }
        );
    }
    let forward = find("bool", "omitted", "forward");
    let reverse = find("bool", "omitted", "reverse");
    let text = |s: &str| json!(["str", s.chars().map(u32::from).collect::<Vec<_>>()]);
    assert_eq!(
        forward["columns"],
        json!([text("x"), text("datetime"), text("stock_id")])
    );
    assert_eq!(
        reverse["columns"],
        json!([text("stock_id"), text("datetime"), text("x")])
    );
    assert_eq!(
        forward["values"][0],
        json!([["bool", true], ["float", "nan"]])
    );
    assert_eq!(
        reverse["values"][2],
        json!([["float", "nan"], ["bool", true]])
    );
    assert_eq!(
        find("bool", "none", "forward")["values"][0],
        json!([["bool", true], ["none"]])
    );
}
